Sources/MapboxCoreNavigation/RouteController.swift

```diff
@@ -321,19 +340,12 @@ open class RouteController: NSObject {
         
         updateIndexes(status: status, progress: routeProgress)
         updateRouteLegProgress(status: status)
-        let willReroute = !userIsOnRoute(snappedLocation, status: status)
-        && (delegate?.router(self, shouldRerouteFrom: snappedLocation)
-            ?? DefaultBehavior.shouldRerouteFromLocation)
         
-        updateSpokenInstructionProgress(status: status, willReRoute: willReroute)
+        updateSpokenInstructionProgress(status: status, willReRoute: isRerouting)
```

Comment:
This is a bit incongruous. If rerouting is already taking place, then it’s too late to send a “will reroute” notification. The application would expect to be able to do things like preempt its own network calls ahead of time.

Reply:
I am not sure to get your point. Delegate's `willRerouteFrom` method is called after `shouldRerouteFrom` which is the earliest notice we get when reroute is incoming. User is still able to cancel this action.
Since NN now takes care on detecting when user is on or off-route and when reroute is beginning, we are setting `isRerouting` flag according to their notifications and user accepting the reroute via `shouldRerouteFrom`.  This is the closest equivalent to previous implementation we can get. Even though we still have `userIsOnRoute`, we cannot reuse it and guarantee that it is the same as a detected reroute event because full reroute detecting logic is now a black box to us.

Reply:
This feedback was only about the timing – that “will reroute” shouldn’t be sent during a reroute, only before it. In the most trivial implementation of KVO lifecycle delegate methods, the “will” method can be called synchronously right after the “should” method without waiting for something else to happen.

I could’ve chosen a clearer hypothetical scenario, but a more relevant consideration is that apparently it’s possible for the navigator to send us location/progress updates while we’re waiting for a new route. If that’s the case, then this would be an unexpectedly late time to tell the application that it will reroute.

Maybe the disconnect is in what “reroute” means to each of us. For me, “will reroute” means “will start calculating a new route”, rather than “will receive a new route”. After all, a UI would reasonably start showing a spinner when the route controller “will reroute” but no sooner.

Reply:
Unfortunately new implementation on NN desynchronizes reporting status and off route updates.`isRerouting` flag is set immediately when we are sure that reroute will happen (right after `shouldRerouteFrom` and `willRerouteFrom` calls), but we cannot guarantee that it will be related to the same status update, where `updateSpokenInstructionProgress` is called. Technically this means that it is possible that we might receive 'extra' status update(s) after reroute was detected, but I assume that NN's status and offroute updates are both tethered to location updates, so there should not be any significant lag between related updates.

Reply:
I think we’ve already seen some instances of this scenario on Android, but I don’t consider it to be any more of a showstopper than it has been on Android. This only affects the behavior of the delegate methods, not their signatures, so it’s possible that a solution to this problem will be backwards-compatible anyways.